parallelize gamma calculation

yunlanli opened the pull request:
# Description

![parmap_rdeepseq](https://user-images.githubusercontent.com/25857014/208030751-e113880b-3971-457a-9a33-58c8c07b7e90.png)

:+1: $\Gamma$ calculation is parallelized
:-1: `toCoflow` is very sequential and prevents us from obtaining speed ups when using multiple cores

yunlanli committed:
parallelize gamma calculation

yunlanli commented:
The first 3 chapters of Marlow's books are insanely helpful! Here's how I understood why using `parMap rdeepseq` works:

- `parMap strat` uses the strategy `strat` to evaluate each element of the list *in parallel*
- our list element is of the form `(cid, getGamma switchLinkRates coflow)`, what we would like is to fully evaluate this expression, and thus the use of `rdeepseq`. Using `rpar` to reduce to WHNF wouldn't help because the list element is already in WHNF, so the computation we want to parallelize - i.e. `getGamma` - wouldn't be parallelized.

Etesam913 commented:
I'll probably review this tomorrow. Sorry I have been busy yesterday working on my Art Hum. Essay. I will try to get some work on this today.

yunlanli commented:
1. has **dataflow dependencies** that could be parallelized by using the `Par` monad, but I doubt would be worth it since the work is very fine and there's probably more overhead from parallelization than speed up. 

```hs
getGamma :: BandwidthTable -> Coflow -> Rational
getGamma bwTbl (Coflow _ _ ingressFlows egressFlows) =
  max (maximum ingressTimes) (maximum egressTimes)
```

2. parallelizing `toCSP`:

- (1) probably worth it since each ingress switch could have up to 1,000s of flows: a decent amount of work
- (2) could maybe use chunking with `parBuffer`: need to experiment with a good chunk size

```hs
toCoflows :: CSP -> CoflowMap
toCoflows csp =
  foldl update Map.empty $ ingressSwitches csp                          -- (1)
  where
    ...
    update :: CoflowMap -> Switch -> CoflowMap
    update currMap switch =
      foldl addFlow currMap $ zip (repeat $ iId switch) (flows switch)  -- (2)
```

yunlanli commented:
some other work that remains to be done:

- [x] output the total running time excluding problem generation for measuring speed up
- [x] add cmdline opt to specify mode of computation: sequential or parallel
- [x] a benchmarking script for measuring speed up and plot a graph of speed-up v.s. # cores
- [x] measuring overhead of parallelization
- [x] (**low priority**) make our program take in parameters to specify the parameters of the coflow scheduling program
- [ ] (**stretch goal**) per-switch coflow scheduling based on the global coflow order computed by the centralized controller

Etesam913 committed:
⏰ Added clock package to measure time of parSebf

➕ Added formatting package to print the timeSpecs Format type

Etesam913 commented:
I added a package to measure the times of the sebf vs parSebf algorithms
sebf was around **7.92** and parSebf was around **4.73**

It also measures the time it takes to print out the results, so it may be possible to reduce the times.

yunlanli committed:
add dep optparse-applicate for args parse

- added option --type/-t to sepcify mode of varys: parallel or
sequential, defaults to parMap (i.e. parallel)

yunlanli committed:
add cmdline options

yunlanli commented:
added cmdline opt:

```sh
❯ stack exec ParVarys-exe -- -h
ParVarys: Parallel Varys Coflow Scheduling Using SEBF

Usage: ParVarys-exe [-t|--type STRING] [-n|--coflows NUMBER]
                    [-i|--ingress NUMBER] [-e|--egress NUMBER]
                    [-s|--min-flow-size NUMBER] [-S|--max-flow-size NUMBER]
                    [-f|--min-switch-flows NUMBER]
                    [-F|--max-switch-flows NUMBER]
                    [-b|--ingress-bandwith NUMBER] [-B|--egress-bandwith NUMBER]
  Generates an offline coflow scheduling problem, and uses the Varys Shortest
  Effective Bottlneck First heuristic to order the coflows.

Available options:
  -h,--help                Show this help text
  -t,--type STRING         Varys mode: seq, parMap (default: "parMap")
  -n,--coflows NUMBER      Number of coflows (default: 4000)
  -i,--ingress NUMBER      Number of ingress switches (default: 1000)
  -e,--egress NUMBER       Number of egress switches (default: 1000)
  -s,--min-flow-size NUMBER
                           Smallest flow size in bytes (default: 0)
  -S,--max-flow-size NUMBER
                           Largest flow size in bytes (default: 1000)
  -f,--min-switch-flows NUMBER
                           Minimum number of flows arriving at an ingress switch
                           (default: 0)
  -F,--max-switch-flows NUMBER
                           Maximum number of flows arriving at an ingress switch
                           (default: 5000)
  -b,--ingress-bandwith NUMBER
                           Ingress bandwidth (Gb/s) of a switch (default: 40)
  -B,--egress-bandwith NUMBER
                           Egress bandwidth (Gb/s) of a switch (default: 40)
```

yunlanli committed:
add benchmark and plot

yunlanli committed:
update plot notebook w/ current parMap data

Etesam913 committed:
📓 Py notebook reads in results from file

🧮 Parallelization overhead is calculated

yunlanli committed:
update benchmark script and plot

- enable multiple runs
- plot average and error bars
- plot speed up graph

yunlanli committed:
initial attempt at parallelize toCoflows

yunlanli commented:
After parallelizing `toCSP`, we achieve almost linear scaling up to 8 HECs!

![image](https://user-images.githubusercontent.com/25857014/208235887-d3a5a5cb-af02-4c4a-bcee-a5edb31b94af.png)

Moreover, SEBF is very memory-intensive, making garbage collection expensive. Setting a large max heap size and long gc cycle can help mitigate this impact. I used a max heap size of 8GB and a garbage collection cycle of 30 seconds, the effect is pronounced!

![v2_N8_default](https://user-images.githubusercontent.com/25857014/208235828-0c15f815-e87d-4809-be91-33f7591f68a4.png)

With ghc runtime options `-H8G -I30`:

![v2_N8_H8G_I30](https://user-images.githubusercontent.com/25857014/208235836-ab7fe428-7e0c-4b27-8a35-fe9e67fa1bd9.png)

yunlanli commented:
some final work that remains to be done:

- [x] is it possible to seed randomRIO so that we can get deterministic problems to solve?
- [x] investigate why our speed-up curve drops off when HECs > 8, what's the bottleneck? Can we eliminate it?
- [x] run some other experiments to further improve speed up, need to brainstorm, here are some that I can think of
  - **not needed** will a more coarse granularity to parallelize work be more efficient? - i.e. chunking + `parBuffer`
  - are there places where evaluating to WHNF is enough, to reduce our memory footprint $\rightarrow$ Lazy Map!
- [ ] tidy up our code
  - run an auto-formatter: line width < 80, make it look good
  - review the code and make sure what we are doing conceptually make sense and look for opportunities to simplify design or make the code more readable (naming, etc)
- [ ] what should we demo? do we need to make the output more readable, or maybe visualize the problem / output using python?

Not so important:

- [ ] run experiments (no gc opts) with different problem parameters, plot the graphs and compare results, to investigate the impact of garbage collection on speed up

yunlanli commented:
The total allocated heap size is about 8G for the default problem, using more cores would cause a bigger heap size due to more threads for GC. When HECs > 10, the total allocated heap size grows beyond 8G, so that GC kicks after `toCSP`, which is why we observed a performance drop.
![v2_N16_H8G_I30](https://user-images.githubusercontent.com/25857014/208253144-107c24cc-5f70-42b4-984c-cecc3e5ad4c4.png)

> The first part of parallelization comes from `parToCSP`, the second part of parallelization comes from parallelizing $\Gamma$ calculation. It seems that the work is too fine that in the end, some cores are idle.

Increasing the max heap size to 10G solves this problem:
![v2_N16_H10G_I30](https://user-images.githubusercontent.com/25857014/208253150-ef515d60-4e37-4eb5-9c27-54aebe62d9d4.png)

And now we see the speed-up curve tapers off, instead of drops off, which is good:
![par_perf_H10G_I30](https://user-images.githubusercontent.com/25857014/208253196-3477c09c-1547-4351-b00f-c6ab2540cb79.png)

yunlanli committed:
ignore python cache and pngs

yunlanli committed:
update benchmark results

On Apple Macbook Pro M1 (16G mem, 10 cores):
  benchmark/benchmark.sh 3 -H10G -I30

yunlanli committed:
add opt to seed the global pseduo-random generator

yunlanli committed:
fix bug in parallel implementation of toCoflow

yunlanli committed:
use lazy IntMap

yunlanli committed:
sparks with finer granularity

yunlanli committed:
use special case of chunkisze=n for n els

yunlanli committed:
add performance benchmarks for final version

Etesam913 committed:
🎨 Added autoformatter named britanny
